lib/gssapi/mech/gss_import_name.c

```diff
@@ -299,3 +297,71 @@ gss_import_name(OM_uint32 *minor_status,
 	gss_release_name(&ms, &rname);
 	return major_status;
 }
+
+static OM_uint32
+_gss_import_composite_export(OM_uint32 *minor_status,
+                             const gss_buffer_t input_name_buffer,
+                             gss_name_t *output_name)
+{
+    OM_uint32 major_status;
+    unsigned char *p = input_name_buffer->value;
+    size_t len = input_name_buffer->length;
+    gss_OID_desc mech_oid;
+    gssapi_mech_interface m;
+    struct _gss_name *name;
+    gss_name_t new_canonical_name;
+    uint16_t oidlen;
+
+    *minor_status = 0;
+    *output_name = 0;
+
+    /*
+     * Make sure that TOK_ID is {4, 2}.
+     */
+    if (len < 4)
+        return GSS_S_BAD_NAME;
+    if (p[0] != 4 || p[1] != 2)
+        return GSS_S_BAD_NAME;
+    p += 2;
+    len -= 2;
+    oidlen = p[0] << 8 | p[1];
+    p += 2;
+    len -= 2;
+
+    if (len < oidlen)
+        return GSS_S_BAD_NAME;
+
+    mech_oid.length = oidlen;
+    mech_oid.elements = p;
+    p += oidlen;
+    len -= oidlen;
+
+    m = __gss_get_mechanism(&mech_oid);
+    if (!m || !m->gm_import_name)
+        return (GSS_S_BAD_MECH);
+
+    /*
+     * Ask the mechanism to import the name.
+     */
+    major_status = m->gm_import_name(minor_status, input_name_buffer,
+                                     GSS_C_NT_COMPOSITE_EXPORT,
+                                     &new_canonical_name);
+    if (major_status != GSS_S_COMPLETE) {
+        _gss_mg_error(m, *minor_status);
+        return major_status;
+    }
+
+    /*
+     * Now we make a new name and mark it as an MN.
+     */
+    name = _gss_create_name(new_canonical_name, m);
```

Comment:
Remind me the rationale behind factoring this out?

Reply:
The token format I chose to use is fairly different from the dumb one in RFC 2743.